### ISSUE TITLE
Can't use sqlite_ecto with latest Phoenix framework

### ISSUE THREAD

**Issue by markokocic** [jazzyb/sqlite_ecto#47]
sqlite_ecto doesn't work with latest ecto version shipped with Phoenix Framework 1.3 due to dependency resolution problems.

**Comment by sprobstCocus** [comment]
Looking at the `ecto` project, it claims that `sqlite_ecto` is not compatible with `ecto` 2.0. `phoenix_ecto` needs `ecto` version `2.0.0-rc`, while `sqlite_ecto` requires version `1.1`.
I don't know how hard it is to update `sqlite_ecto` to support the latest version of `ecto` .

**Comment by HughGrant** [comment]
same problem here. not just `ecto`2, but also with `posion`2.

**Comment by hayesgm** [comment]
Does this project have a plan or roadmap to support Ecto 2.0? It's like a hard requirement for long-term use of this project.

**Comment by kentor** [comment]
Hmm yeah.. this is really hurting adoption / learning of ecto/phoenix. Not really interested in setting up an SQL server for testing purposes.

**Comment by philipgiuliani** [comment]
Yeah and for nerves-project :( Dont want to install any DB on embedded.

**Comment by hayesgm** [comment]
How much effort will be it to make this project Ecto 2.0 compatible? Probably worth the effort to tag out the work effort and just move on it. I think this is too important to just ignore.

**Comment by gallexme** [comment]
any updates?

**Comment by obmarg** [comment]
I have been working on adding ecto 2 support over the weekend. Got the basics working already, just need to fix a bunch of integration tests.  Hope to have a PR ready in the next few days.

**Comment by philipgiuliani** [comment]
Thank you very much for your work @obmarg! You already have it on git so i can try it out?

**Comment by obmarg** [comment]
@philipgiuliani I've pushed up a WIP commit to the `feature/ecto-2.0` branch of my fork: https://github.com/obmarg/sqlite_ecto/tree/feature/ecto-2.0

I can't guarantee it's particularly usable right now: the project I've been testing it against has fairly minimal usage of ecto, and the integration tests are failing before they even start.  Feel free to give it a try though.

**Comment by alakra** [comment]
@jazzyb any thoughts on this?

**Comment by philipgiuliani** [comment]
@obmarg Could you also update Poison to 2.0? The sqlite_ecto requires `~> 1.0.0` and my project requires `~> 2.0.0`. Ecto itself requires `~> 1.5 or ~2.0`.

So far it seems to work fine. I have tested to create migrations, migrate and insert/read data from tables. But without associations yet.

I also noticed that `:sqlitex, :db_connection` have to be added in applications. Otherwise exrm will not add them to the release.

**Comment by fishcakez** [comment]
I think to use SQLite with Ecto 2 we will need to switch to a port_driver based API, instead of NIF because DBConnection is designed around multi process port access and wants to be able to close the the "connection" while another process is running a query and to close the "connection" when the DBConnection process exits. This would mean replacing `sqlitex` with a different SQLite implementation.

https://github.com/alexeyr/erlang-sqlite3 seems like a reasonable candidate. Likely we would want to fork it and replace rebar with mix, and the Erlang code with Elixir.

**Comment by obmarg** [comment]
@fishcakez I'm not sure I follow, how does Sqlitex being implemented with NIFs prevent DBConnection from closing "connections"?  

I've got an [incomplete DBConnection implementation for Sqlitex](https://github.com/obmarg/sqlite_ecto/blob/feature/ecto-2.0/lib/sqlite_ecto/protocol.ex), is there something obvious I'm missing in there?

**Comment by fishcakez** [comment]
@obmarg I am concerned about the Ecto 2.0 style timeouts and error handling. Ecto 2.0 timeouts require the "connection" to be closed by DBConnection while the caller is running queries, which interrupts the caller. I am unsure how that can be achieved with a NIF. In the code you linked it looks like this doesn't happen but please correct me if I am wrong. With a port_driver we can close the port. With Ecto 2.0 timeouts are per ownership and per transaction - different to Ecto 1.0.
